compiler/rustc_typeck/src/coherence/orphan.rs

```diff
@@ -277,42 +277,78 @@ fn emit_orphan_check_error<'tcx>(
+                            let mut err = struct_span_err!(
+                                tcx.sess,
+                                sp,
+                                E0210,
+                                "type projection `{}` must be covered by another type \
+                                when it appears before the first local type (`{}`)",
+                                param_ty,
+                                local_type
+                            );
+                            err.span_label(
+                                sp,
+                                format!(
+                                    "type projection `{}` must be covered by another type \
+                                    when it appears before the first local type (`{}`)",
+                                    param_ty, local_type
+                                ),
+                            );
+                            err
+                        }
+                        _ => {
+                            let mut err = struct_span_err!(
+                                tcx.sess,
+                                sp,
+                                E0210,
+                                "type parameter `{}` must be covered by another type \
+                                when it appears before the first local type (`{}`)",
+                                param_ty,
+                                local_type
+                            );
+                            err.span_label(
+                                sp,
+                                format!(
+                                    "type parameter `{}` must be covered by another type \
+                                    when it appears before the first local type (`{}`)",
+                                    param_ty, local_type
+                                ),
+                            );
+                            err
+                        }
+                    };
+
+                    err.note(
+                        "implementing a foreign trait is only possible if at \
                         least one of the types for which it is implemented is local, \
-                        and no uncovered type parameters appear before that first \
+                        and no uncovered type parameters or projections appear before that first \
                         local type",
-                )
-                .note(
-                    "in this case, 'before' refers to the following order: \
+                    )
+                    .note(
+                        "in this case, 'before' refers to the following order: \
                         `impl<..> ForeignTrait<T1, ..., Tn> for T0`, \
                         where `T0` is the first and `Tn` is the last",
-                )
-                .emit(),
+                    )
+                    .emit()
+                }
                 None => struct_span_err!(
```

Comment:
this branch can be hit by
```rust

extern crate coherence_projection_assoc as lib;
use lib::Foreign;

trait Id {
    type Assoc;
}

impl<T> Id for T {
    type Assoc = T;
}

impl<T> Foreign<(), Vec<T>> for <T as Id>::Assoc {
//~^ ERROR E0210
    type Assoc = usize;
}

fn main() {}
```

Reply:
What would you advise for this? Should I duplicate the match looking for ty::Projection in the None branch too?